torchsynth/module.py

```diff
@@ -541,22 +540,22 @@ def _forward(self, audio_in: Signal, control_in: Signal) -> Signal:
 
 class Noise(SynthModule):
     """
-    Generates white noise that is the same length as the buffer
+    Generates white noise that is the same length as the buffer.
+
+    A batch size number of noise samples are pre-computed for performance.
+    A seed for the noise generator is required. If you use multiple of
+    these in a synth then choose different seeds for each instance.
     """
 
-    def __init__(self, synthglobals: SynthGlobals, **kwargs):
+    def __init__(self, synthglobals: SynthGlobals, seed: int, **kwargs):
         super().__init__(synthglobals, **kwargs)
-        self.register_buffer(
-            "noise",
-            torch.empty((self.batch_size, self.buffer_size), device=self.device),
-        )
+
+        # Create a prime number of noise samples, we will cycle through these
+        torch.random.manual_seed(seed)
```

Comment:
Wait no, we can't do this. There two pernicious anti-patterns in RNGs that people fall into:
1) Not using deterministic seeds when they should. It makes it harder to debug stuff retroactively. We've fixed this.
2) Changing the global seed in an inner loop.

Let's say someone calling our library is using torch.seed to randomize hyperparameters. BOOM now they are getting the same hyperparameters because we mucked with the global seed.

I would comb the code for any time we do random seeding (in random.seed, np.random.seed, pytorch_lightning.seed_everything, and torch.manual_seed) to make sure we don't mess with downstream clients of our code.

Here you probably want to do this: https://pytorch.org/docs/stable/random.html#torch.random.fork_rng

I can't really figure out how it works so actually....

It might be easier and safer to do this:

oldrngstate = torch.random.get_rng_state()
seed, do some random stuff, etc.
torch.random.set_rng_state(oldrngstate)

Reply:
Good catch. I think we can avoid the global random number generator altogether and just create new ones that we can seed for these sort of operations.